src/bluesky/simulators.py

```diff
@@ -102,5 +117,285 @@ async def check_limits_async(plan):
+            in order.
+        """
+        if not docs:
+            if not document_name or not document:
+                raise ValueError("No document sequence specified and document or document_name not specified")
+            docs = [[(document_name, document)]]
+        it = iter(docs)
+
+        def handle_command(_):
+            for name, doc in next(it):
+                self.fire_callback(name, doc)
+
+        self.add_handler(command, handle_command, msg_filter)
+
+    def fire_callback(self, document_name, document) -> None:
+        """Fire all the callbacks registered for this document type in order to simulate something happening.
+
+        Parameters
+        ----------
+        document_name
+            document name as defined in the Bluesky Message Protocol 'subscribe' call,
+            all subscribers filtering on this document name will be called
+        document
+            the document to send
+        """
+        for callback_func, callback_docname in self.callbacks.values():
+            if callback_docname == "all" or callback_docname == document_name:
+                callback_func(document_name, document)
+
+    def simulate_plan(self, gen: Generator[Msg, Any, Any]) -> List[Msg]:
+        """Simulate the RunEngine executing the plan.
+
+        After executing the plan return_value is populated with the return value of the plan.
+
+        Parameters
+        ----------
+        gen
+            the generator function that executes the plan
+
+        Returns
+        -------
+        list[Msg]
+            list of the messages generated by the plan
+        """
+        messages = []
+        send_value = None
+        try:
+            while msg := gen.send(send_value):
+                send_value = None
+                messages.append(msg)
+                LOGGER.debug(f"<{msg}")
+                if handler := next((h for h in self.message_handlers if h.predicate(msg)), None):
+                    send_value = handler.runnable(msg)
+
+                if send_value:
+                    LOGGER.debug(f">send {send_value}")
+        except StopIteration as e:
+            self.return_value = e.value
+        return messages
+
+    def _add_callback(self, msg_args):
+        self.callbacks[self.next_callback_token] = msg_args
+        self.next_callback_token += 1
+
+
+def assert_message_and_return_remaining(
+    messages: List[Msg],
+    predicate: Callable[[Msg], bool],
+    group: Optional[str] = None,
+):
+    """Find the next message matching the predicate, assert that we found it.
+
+    Returns
+    -------
+    list[Msg]
+        All the remaining messages starting from the matched message which is included as a
+        convenience to capture information that may be used in subsequent calls.
+    """
+    indices = [
+        i
+        for i in range(len(messages))
+        if (not group or (messages[i].kwargs and messages[i].kwargs.get("group") == group))
```

Comment:
I've changed the code to handle unnamed groups. We were writing a number of tests which were asserting various operations performed in groups that were subsequently`wait`ed and it seemed like this would be useful.